app/views/govuk_component/task_list.raw.html.erb

```diff
@@ -17,70 +19,132 @@
+                <span class="visuallyhidden">Step</span> <%= group_index + 1 %>
+              </span>
+            </span>
           </span>
+
           <% group.each_with_index do |step, step_index| %>
             <%
               step_count += 1
               id = step[:id] || step[:title].downcase.tr(" ","-")
+
+              logic = "and"
+              logic = step[:logic] if ["and", "or"].include? step[:logic]
             %>
             <div
-              class="pub-c-task-list__step js-step"
+              class="pub-c-task-list__step js-step <% if step[:optional] %>pub-c-task-list__step--optional<% end %>"
               id="<%= id %>"
               data-track-count="tasklistSection"
-              <% if step_count == open_step %>data-open<% end %>
+              <% if step_count == show_step %>data-show<% end %>
               >
+              <% if step_index > 0 and logic.length %>
+                <div class="pub-c-task-list__circle pub-c-task-list__circle--logic">
+                  <span class="pub-c-task-list__circle-inner">
+                    <span class="pub-c-task-list__circle-background">
+                      <%= logic %>
+                    </span>
+                  </span>
+                </div>
+              <% end %>
+
               <div class="pub-c-task-list__header js-toggle-panel" data-position="<%= "#{group_index + 1}.#{step_index + 1}" %>">
                 <h<%= heading_level %> class="pub-c-task-list__title js-step-title">
                   <%= step[:title] %>
                 </h<%= heading_level %>>
               </div>
 
               <div class="pub-c-task-list__panel js-panel" id="step-panel-<%= id %>-<%= step_index + 1 %>">
-                <div class="pub-c-task-list__panel-content">
-                  <%= step[:panel].html_safe if step[:panel] %>
+                <%
+                  in_substep = false
+                  link_index = 0
+                %>
+                <% step[:contents].each do |element| %>
+                  <% if element[:type] == 'paragraph' %>
+                    <p class="pub-c-task-list__paragraph"><%= element[:text] %></p>
 
-                  <% step[:panel_descriptions].to_a.each do |panel_description| %>
-                    <p class="pub-c-task-list__panel-description">
-                      <%= panel_description %>
-                    </p>
-                  <% end %>
-                </div>
+                  <% elsif element[:type] == 'list' %>
+                    <%
+                      list_style = ''
+                      list_element = 'ol'
+
+                      if element[:style] == 'required'
+                        list_style = 'pub-c-task-list__links--required'
+                      elsif element[:style] == 'choice'
+                        list_style = 'pub-c-task-list__links--choice'
+                        list_element = 'ul'
+                      end
+                    %>
+                    <<%= list_element %>
```

Comment:
Will content_tag let me do the complex stuff needed for the list items within this list? I'm worried even if it will it'll make the code even more complex.

Reply:
Yeah the syntax just takes an arbitrary block of content:

```erb
<%= content_tag(
  list_element,
  class: "pub-c-task-list__links #{list_style},
  data: {
    length: element[:contents].length
  }) do %>
  <% element[:contents].each do |link| %>
    ...
  <% end %>
<% end %>
```